### PR TITLE
Feature/data sync

### PR DESCRIPTION
**Pull request by smousa**
https://github.com/control-center/serviced/pull/1743

### CODE REVIEW

File: scheduler/localsync.go [hunk 1]
```diff
@@ -10,83 +10,87 @@
...
-				continue retry
-			} else if err := zkservice.SyncHosts(conn, hosts); err != nil {
-				glog.Errorf("Could not do a local sync of hosts: %s", err)
-				wait = time.After(minWait)
-				continue retry
-			}
+// Purge performs the synchronization for services, hosts, pools, and virtual
+// ip.
+// Implements utils.TTL
+func (sync *LocalSync) Purge(age time.Duration) (time.Duration, error) {
```
Comment (wwwtyro) [inline scheduler/localsync.go]: Are SyncVirtualIPs, SyncHosts, and SyncServices idempotent? i.e., if something goes wrong in the loop starting on line 69 and this function exits prematurely, is it going to do something screwy to the data when it runs again?

Comment (smousa) [inline scheduler/localsync.go]: no

File: scheduler/remotesync.go [hunk 2]
```diff
@@ -14,180 +14,114 @@
...
+		glog.Errorf("Could not look up remote services for pool %s: %s", sync.poolID, err)
+		return 0, err
+	}
 
-	// Add the service listener
-	poolSync.AddListener(func(id string) zzk.Listener {
-		return zkservice.NewServiceSynchronizer(s, id)
-	})
+	// build the datamap
+	datamap := make(map[string]interface{})
```
Comment (wwwtyro) [inline scheduler/remotesync.go]: I know utils.Sync takes a map[string]interface{}, but does this need to be generic? Will Go not compile this if it's a map[string]*service.Service?

Comment (smousa) [inline scheduler/remotesync.go]: correct

File: scheduler/remotesync.go [hunk 3]
```diff
@@ -14,180 +14,114 @@
...
+		return svc, nil
 	}
-
-	return s.facade.UpdateHost(datastore.Get(), host)
-}
-
-func (s *scheduler) RemoveHost(id string) error {
-	return s.facade.RemoveHost(datastore.Get(), id)
-}
+	return nil, errors.New("invalid type")
```
Comment (wwwtyro) [inline scheduler/remotesync.go]: Just nitpicking, but maybe use a more expressive string here? I dunno, "Could not convert data to Service object" or something?

Comment (smousa) [inline scheduler/remotesync.go]: ok

File: zzk/service/hostsync.go [hunk 4]
```diff
@@ -10,82 +10,78 @@
...
+	poolID string
 }
 
-// HostSynchronizer is the synchronizer for Host data
-type HostSynchronizer struct {
-	handler HostSyncHandler
-	poolID  string
+// SyncHosts performs the host synchronization
+func SyncHosts(conn client.Connection, poolID string, hosts []host.Host) error {
+	hostmap := make(map[string]interface{})
```
Comment (wwwtyro) [inline zzk/service/hostsync.go]: Same question here, can this be a map[string]*host.Host?

File: zzk/service/hostsync.go [hunk 5]
```diff
@@ -10,82 +10,78 @@
...
+func (sync *HostSync) Delete(id string) error {
+	return sync.conn.Delete(poolpath(sync.poolID, hostpath(id)))
 }
+
+// convert transforms the generic object data into its proper type
+func (sync *HostSync) convert(data interface{}) (string, *host.Host, error) {
+	if host, ok := data.(*host.Host); ok {
+		return poolpath(sync.poolID, hostpath(host.ID)), host, nil
+	}
+	return "", nil, errors.New("invalid type")
```
Comment (wwwtyro) [inline zzk/service/hostsync.go]: Same nitpick about the error string.

File: zzk/service/hostsync.go (hunk 4 above)
Comment (smousa) [inline zzk/service/hostsync.go]: no

File: zzk/service/poolsync.go [hunk 6]
```diff
@@ -14,81 +14,73 @@
...
+	conn client.Connection // the global client connector
 }
 
-// NewPoolSynchronizer initializes a new Synchronizer
-func NewPoolSynchronizer(handler PoolSyncHandler, getConnection zzk.GetConnection) *zzk.Synchronizer {
-	pSync := &PoolSynchronizer{handler, getConnection}
-	return zzk.NewSynchronizer(pSync)
+// SyncPools performs the pool synchronization
+func SyncPools(conn client.Connection, pools []pool.ResourcePool) error {
+	poolmap := make(map[string]interface{})
```
Comment (wwwtyro) [inline zzk/service/poolsync.go]: Same question about generic type.

Comment (smousa) [inline zzk/service/poolsync.go]: again, no

File: zzk/service/poolsync.go [hunk 7]
```diff
@@ -14,81 +14,73 @@
...
+func (sync *PoolSync) Delete(id string) error {
+	return sync.conn.Delete(poolpath(id))
 }
+
+// convert transforms the generic object data into its proper type
+func (sync *PoolSync) convert(data interface{}) (string, *pool.ResourcePool, error) {
+	if pool, ok := data.(*pool.ResourcePool); ok {
+		return poolpath(pool.ID), pool, nil
+	}
+	return "", nil, errors.New("invalid type")
```
Comment (wwwtyro) [inline zzk/service/poolsync.go]: Same nitpick.

File: zzk/service/servicesync.go [hunk 8]
```diff
@@ -10,82 +10,78 @@
...
+func (sync *ServiceSync) Delete(id string) error {
+	return sync.conn.Delete(poolpath(sync.poolID, servicepath(id)))
 }
+
+// convert transforms the generic object data into its proper type
+func (sync *ServiceSync) convert(data interface{}) (string, *service.Service, error) {
+	if service, ok := data.(*service.Service); ok {
+		return poolpath(sync.poolID, servicepath(service.ID)), service, nil
+	}
+	return "", nil, errors.New("invalid type")
```
Comment (wwwtyro) [inline zzk/service/servicesync.go]: Same nitpick.

File: zzk/virtualips/vipsync.go [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+func (sync *VirtualIPSync) Delete(id string) error {
+	return sync.conn.Delete(poolpath(sync.poolID, vippath(id)))
+}
+
+// convert transforms the generic object data into its proper type
+func (sync *VirtualIPSync) convert(data interface{}) (string, *pool.VirtualIP, error) {
+	if vip, ok := data.(*pool.VirtualIP); ok {
+		return poolpath(sync.poolID, vippath(vip.IP)), vip, nil
+	}
+	return "", nil, errors.New("invalid type")
```
Comment (wwwtyro) [inline zzk/virtualips/vipsync.go]: Same nitpick.

File: scheduler/localsync.go (hunk 1 above)
Comment (wwwtyro) [inline scheduler/localsync.go]: No to which question?

Comment (smousa) [inline scheduler/localsync.go]: it is not going to affect the data.  It will exit and restart after 30 seconds

**Pull request merged**